src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,158 @@
+            'created_at' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+        ]);
+    }
+
+    public function get(): ?array
+    {
+        if ($this->debug) {
+            $this->setup();
+        }
+
+        $this->driverConnection->beginTransaction();
+        try {
+            $query = $this->driverConnection->createQueryBuilder()
+                ->select('m.*')
+                ->from($this->tableName, 'm')
+                ->where('m.status = :status')
+                ->orderBy('id', 'ASC')
+                ->setMaxResults(1);
+
+            // use SELECT ... FOR UPDATE to lock table
+            $doctrineEnvelope = $this->driverConnection->executeQuery(
+                $query->getSQL().' '.$this->driverConnection->getDatabasePlatform()->getWriteLockSQL(),
+                [
+                    ':status' => self::STATUS_PENDING,
+                ]
+            )->fetch();
+
+            if (false === $doctrineEnvelope) {
+                $this->driverConnection->commit();
+
+                return null;
+            }
+
+            $doctrineEnvelope['headers'] = \json_decode($doctrineEnvelope['headers'], true);
+
+            $this->driverConnection->createQueryBuilder()
+                ->update($this->tableName)
+                ->set('status', ':delivered_status')
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id')
+                ->setParameter('delivered_status', self::STATUS_DELIVERED)
+                ->setParameter('delivered_at', (new \DateTime('now'))->format('Y-m-d H:i:s'))
+                ->setParameter('id', $doctrineEnvelope['id'])
+                ->execute();
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Exception $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->tableName, ['id' => $id]) > 0;
+    }
+
+    public function nack($id): bool
+    {
+        return $this->driverConnection->update($this->tableName, ['status' => self::STATUS_PENDING], ['id' => $id]) > 0;
+    }
+
+    public function requeueMessages()
+    {
+        $redeliverLimit = new \DateTime('now');
+        $redeliverLimit->modify(sprintf('-%d second', $this->redeliverTimeout));
+        $this->driverConnection->createQueryBuilder()
+            ->update($this->tableName)
+            ->set('status', ':pending_status')
+            ->set('delivered_at', ':delivered_at')
+            ->where('status = :delivered_status')
+            ->andWhere('delivered_at < :redeliver_limit')
+            ->setParameter('pending_status', self::STATUS_PENDING)
+            ->setParameter('delivered_at', null)
+            ->setParameter('delivered_status', self::STATUS_DELIVERED)
+            ->setParameter('redeliver_limit', $redeliverLimit->format('Y-m-d H:i:s'))
```

Comment:
Is `delivered_at` used anywhere? Or is it just informational?

Also, one requirement (for me) is that we support: delayed messages. So, we put them in the table, but with a date column that says when it's ok to actually fetch them.

Thanks!

Reply:
The delivered_at is used to requeue messages that are not ack (e.g: the handler failed and there was a Database connection error, a server crash, ...). 

I've update the PR details to explain more how this transport works.

For the delaying of message. How can we configure this during the dispatch ? via a Stamp ? It must be nice to have this for all transport. For the moment I can set the `available_at` to `now` when inserting the message into the table.